cmake/patches/dawn/dawn.patch

```diff
@@ -79,3 +79,111 @@ index 0037d83276..6372c4ee77 100644
    tint_lang_wgsl_program
    tint_lang_wgsl_sem
    tint_lang_wgsl_writer_ir_to_program
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index bcb91ba106..4e54940c50 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -188,6 +188,8 @@ option(DAWN_ENABLE_PIC "Build with Position-Independent-Code enabled" OFF)
+ option(DAWN_EMIT_COVERAGE "Emit code coverage information" OFF)
+ set_if_not_defined(LLVM_SOURCE_DIR "${Dawn_LLVM_SOURCE_DIR}" "Directory to an LLVM source checkout. Required to build turbo-cov")
+
++option(DAWN_CROSSCOMPILING_WIN "Support windows platform crosscompiling, mainly for build arm/aarch64 target on x64 platform" OFF)
++
+ message(STATUS "Dawn build with asserts in all configurations: ${DAWN_ALWAYS_ASSERT}")
+ message(STATUS "Dawn build Wayland support: ${DAWN_USE_WAYLAND}")
+ message(STATUS "Dawn build X11 support: ${DAWN_USE_X11}")
+@@ -377,6 +380,11 @@ target_include_directories(dawn_internal_config INTERFACE
+ )
+ target_link_libraries(dawn_internal_config INTERFACE dawn_public_config)
+
++# Disable CMAKE_CROSSCOMPING because we don't have full support
++if(CMAKE_CROSSCOMPILING)
++  set(CMAKE_CROSSCOMPILING OFF)
++endif()
++
+ ################################################################################
+ # Include utility CMake modules
+ ################################################################################
+diff --git a/third_party/CMakeLists.txt b/third_party/CMakeLists.txt
+index bcb91ba106..4e54940c50 100644
+--- a/third_party/CMakeLists.txt
++++ b/third_party/CMakeLists.txt
+@@ -194,6 +194,38 @@ function(get_all_targets_recursive targets dir)
+ endfunction()
+
+ function(AddSubdirectoryDXC)
++    if (DAWN_CROSSCOMPILING_WIN)
++        set(FOLDER_TO_CHECK ${CMAKE_BINARY_DIR}/native-tools)
++        if(EXISTS "${FOLDER_TO_CHECK}")
++             # Find llvm-tblgen and clang-tblgen executable file in the folder
++            find_program(CLANG_TBLGEN NAMES clang-tblgen clang-tblgen.exe)
++            find_program(LLVM_TBLGEN NAMES llvm_tblgen llvm-tblgen.exe)
++        endif()
++
++        # Config clang-tblgen and llvm-tblgen tools with host cpu arch and build them during cmake configuration progress to support future crosscompiling.
++        if (NOT EXISTS "${FOLDER_TO_CHECK}")
++            file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/native-tools)
++        endif()
++
++        if (NOT CLANG_TBLGEN OR NOT LLVM_TBLGEN)
```

Comment:
This part is complicated. I think we may ask the user to provide a prebuilt binary instead.  Like what we do now for protoc.exe .

Reply:
When we build onnxruntime on Windows, we also pass in a lot of other cmake parameters like "-A Win32 --toolset 14.40". These flags cannot be handled here.

Reply:
>This part is complicated. I think we may ask the user to provide a prebuilt binary instead. Like what we do now for protoc.exe 

It would be ideal that user could provided the pre-baked tool for this. 

>When we build onnxruntime on Windows, we also pass in a lot of other cmake parameters like "-A Win32 --toolset 14.40". These flags cannot be handled here.
My understanding is that  in crosscompilation situation, the llvm-talgen and clang-tblgen are built in default behaviour? (Because we don't have extra params to define the behaviour)

Reply:
> the released llvm package contains llvm-tblgen and clang-tblgen tools and that doesn't work with dxc compilation.

No, it would not work. We need the special DXC one.
I was just thinking, if we could achieve the goal without patching dawn, I'd prefer not doing that.

Reply:
I noticed that with cmake flag `CMAKE_CROSSCOMPILING`, onnxruntime generated `NATIVE` folder in build-dir and configs with host-cpu-os info (e.g. Generated visual studio project files with x64 arch while target platform is arm64 by setting -A flag). Several third_party repo works but dawn doesn't. Do you think it is possible that we could rely on this?

Another thoughts is that we might host dxc as a third_party repo and compile tools from it when cross-compilation is required.